phidgets_spatial/src/spatial_ros_i.cpp

```diff
@@ -0,0 +1,295 @@
+{
+    ROS_INFO("Starting Phidgets SPATIAL");
+
+    ROS_INFO("Opening spatial");
+    int serial_num;
+    if (!nh_private_.getParam("serial", serial_num))
+    {
+        serial_num = -1;  // default open any device
+    }
+    int hub_port;
+    if (!nh_private.getParam("hub_port", hub_port))
+    {
+        hub_port = 0;  // only used if the device is on a VINT hub_port
+    }
+    if (!nh_private_.getParam("frame_id", frame_id_))
+    {
+        // As specified in http://www.ros.org/reps/rep-0145.html
+        frame_id_ = "imu_link";
+    }
+    if (!nh_private_.getParam("linear_acceleration_stdev",
+                              linear_acceleration_stdev_))
+    {
+        linear_acceleration_stdev_ = 300.0 * 1e-6 * G;  // 300 ug as per manual
+    }
+    if (!nh_private_.getParam("angular_velocity_stdev",
+                              angular_velocity_stdev_))
+    {
+        angular_velocity_stdev_ =
+            0.02 * (M_PI / 180.0);  // 0.02 deg/s resolution, as per manual
+    }
+    if (!nh_private_.getParam("magnetic_field_stdev", magnetic_field_stdev_))
+    {
+        magnetic_field_stdev_ =
+            0.095 * (M_PI / 180.0);  // 0.095°/s as per manual
+    }
+    int data_interval_ms;
+    if (!nh_private.getParam("data_interval_ms", data_interval_ms))
+    {
+        data_interval_ms = 250;
+    }
+    if (!nh_private.getParam("publish_rate", publish_rate_))
+    {
+        publish_rate_ = 5;
+    }
+
+    // compass correction params (see
+    // http://www.phidgets.com/docs/1044_User_Guide)
+    double cc_mag_field;
+    double cc_offset0;
+    double cc_offset1;
+    double cc_offset2;
+    double cc_gain0;
+    double cc_gain1;
+    double cc_gain2;
+    double cc_T0;
+    double cc_T1;
+    double cc_T2;
+    double cc_T3;
+    double cc_T4;
+    double cc_T5;
+
+    bool has_compass_params =
+        nh_private_.getParam("cc_mag_field", cc_mag_field) &&
+        nh_private_.getParam("cc_offset0", cc_offset0) &&
+        nh_private_.getParam("cc_offset1", cc_offset1) &&
+        nh_private_.getParam("cc_offset2", cc_offset2) &&
+        nh_private_.getParam("cc_gain0", cc_gain0) &&
+        nh_private_.getParam("cc_gain1", cc_gain1) &&
+        nh_private_.getParam("cc_gain2", cc_gain2) &&
+        nh_private_.getParam("cc_t0", cc_T0) &&
+        nh_private_.getParam("cc_t1", cc_T1) &&
+        nh_private_.getParam("cc_t2", cc_T2) &&
+        nh_private_.getParam("cc_t3", cc_T3) &&
+        nh_private_.getParam("cc_t4", cc_T4) &&
+        nh_private_.getParam("cc_t5", cc_T5);
+
+    ROS_INFO(
+        "Waiting for Phidgets Spatial serial %d, hub port %d to be "
+        "attached...",
+        serial_num, hub_port);
+
+    // We take the mutex here and don't unlock until the end of the constructor
+    // to prevent a callback from trying to use the publisher before we are
+    // finished setting up.
+    std::lock_guard<std::mutex> lock(spatial_mutex_);
+
+    spatial_ = std::make_unique<Spatial>(
```

Comment:
Oh, and the "Waiting to be attached" is a bit misleading and should be rephrased, since we don't really wait any more (only 1 second, instead of indefinitely).

Reply:
Commit dfdc7e637023e4886235c9f91de66a8c8d6dcadd updates the wording to be `Connecting to ...`, which I think is a little clearer.

The only thing I can think to do about the failure case for nodelets is to put a `try/catch` block around the connection, and print out the error when it fails, and re-raise.  I'll implement that for all of the nodes.